ttools/ttools-2.0beta2/docs/dast.1
.\" -*- nroff -*-
.\" Copyright 1995 by Yury Shevchuk.  All Rights Reserved.
.\" This file may be copied under the terms of GNU General Public License.
.\"
.Id $Id: dast.1,v 1.1.1.1 1995/12/22 12:24:39 sizif Exp $
.TH dast 1 "$Date: 1995/12/22 12:24:39 $" "botik" "TTOOLS"
.\"-------------------------------
.SH NAME
dast \- binary image disassembler for transputers
("Dizassembler Binarnogo Obraza")
.\"-------------------------------
.SH SYNOPSIS
.B dast
.RB [ options ]
.RI [ FILE ]
.\"-------------------------------
.SH DESCRIPTION
.B dast
allows one to view a binary
.I FILE
as transputer instructions.  If
.I FILE
is not specified,
.B dast
takes input from stdin.  The input might be transputer memory dump,
or an assembled program; however, for the latter consider
.RB `` "litdump --disassemble" ''.

The output of
.B dast
looks like this:

.nf
$ dast pushpull.image
[ ... ]
0x00000025: G        47                ldc 7
0x00000026: %.       25 10             ldlp 80 (0x50)
0x00000028: .        fd                startp
0x00000029: $O       24 4f             ldc 79 (0x4f)
0x0000002b: *.       2a 10             ldlp 160 (0xa0)
0x0000002d: .        fd                startp
0x0000002e: !.       21 f5             stopp
0x00000030: /*K      2f 2a 4b          ldc 4011 (0xfab)
[ ... ]
$
.fi

Fourth column of every line contains a disassembled command, the third
column contains hex dump of the memory occupied by the command
(leftmost byte has the lowest address), the second is the ASCII dump
of the same memory area (nonprintable charactes printed as '.'), and
the first column is the memory address of the command, provided you
have specified the correct origin of the memory image using the
.B -O
option.
.\"-------------------------------
.SH OPTIONS
.\"-------------------------------
.TP
.B --help
show usage summary
.\"-------------------------------
.TP
.BI -o " F\fR," " --output" " F"
output disassembled code to file
.I F\fR.
Defaults to
.B stdout\fR.
.\"-------------------------------
.TP
.BI -s " N\fR," " --skip" " N"
skip first
.I N
bytes of the input; i.e. start disassembling at position
.I N\fR.
.\"-------------------------------
.TP
.BI -O " A\fR," " --origin" " A"
assume the image being disassebled has been taken from the address
.I A
in transputer memory.
.\"-------------------------------
.SH BUGS
If you find any, please submit a description of the bug and the
data the bug exposes on to bug-ttools@botik.ru.
.SH "SEE ALSO"
.BR litdump (1)
.\"-------------------------------
.SH AUTHOR
.B dast
is written by Yury Shevchuk (sizif@botik.ru)
.\"-------------------------------


ttools/ttools-2.0beta2/dast/dast.c
/* $Id: dast.c,v 1.1.1.1 1995/12/22 12:24:39 sizif Exp $ */

/*
 * This file is part of TTOOLS package.
 * Copyright (c) 1995 Yury Shevchuk <sizif@botik.ru>
 */

#include "shared.h"
#include "disassemble.h"

char * infile_n = "stdin";

static char *myname = "litdump";

static char *option_list = 
#undef DEFINE_OPTION
#define DEFINE_OPTION(name, has_arg, flag, val, descr) descr
#include "options.h"
;

static struct option opttab[] = {
#undef DEFINE_OPTION
#define DEFINE_OPTION(name, has_arg, flag, val, x) {name, has_arg, flag, val},
#include "options.h"
};

enum usage_flavor {
    USAGE_GENERAL,
    USAGE_OPTIONS,
};

void
usage (char *errstr, enum usage_flavor flavor, int exit_code)
{
    if (errstr)
	fprintf (stderr, "%s", errstr);

    switch (flavor) {
    case USAGE_GENERAL:
	fprintf (stderr, "This is dast version %s\n", version_string);
	fprintf (stderr, "Usage: %s [options] [BINFILE]\n", myname);
	fprintf (stderr, "%s --help for options list\n", myname);
	break;

    case USAGE_OPTIONS:
	fprintf (stderr, "This is dast version %s\n", version_string);
	fprintf (stderr, "Usage: %s [options] [BINFILE]\n", myname);
	fprintf (stderr, "Options are:\n");
	fprintf (stderr, "%s", option_list);
	fprintf (stderr, "\n");
	break;

    default:
	abort ();
    }

    exit (exit_code);
}

int
main (int argc, char **argv)
{
    static int dummy = 0;
    unsigned long skip = 0;
    unsigned long point = 0;
    char *p;
    int c;
    int i;
    char buf[BUFSIZ + CMD_MAXLEN];
    int bufcount;
    unsigned char *bufp;

    /* Find out how we are named on this system */
    if ((myname = strrchr (argv[0], '/') + 1) == NULL + 1)
	myname = argv[0];

    /* Parse options */
    while ((c = getopt_long (argc, argv, "o:s:O:",
			     opttab, &dummy)) != -1) {
	switch (c) {
	case '?':		/* an illegal option */
	    usage ("", USAGE_GENERAL, 1);

	case 301:		/* "help" */
	    usage ("", USAGE_OPTIONS, 0);

	case 'o':
	    if (! freopen (optarg, "w", stdout)) {
		perror (optarg);
		exit (1);
	    }
	    break;

	case 's':
	    skip = strtoul (optarg, &p, 0);
	    if (*p != '\0') {
		fprintf (stderr, "Bad argument to --skip: `%s'\n", optarg);
		exit (1);
	    }
	    break;

	case 'O':
	    point = strtoul (optarg, &p, 0);
	    if (*p != '\0') {
		fprintf (stderr, "Bad argument to --origin: `%s'\n", optarg);
		exit (1);
	    }
	    break;

	default:
	    abort ();
	}
    }

    if (optind < argc) {
	infile_n = argv[optind];
	if (! freopen (infile_n, "rb", stdin)) {
	    perror (optarg);
	    exit (1);
	}
	if (++optind < argc)
	    usage ("Too many input files given\n", USAGE_GENERAL, 1);
    }

    
    if (skip) {
	if (fseek (stdin, skip, 1) == -1) {
	    /* should be a pipe... skip by read */
	    for ( ; skip > 0; skip -= sizeof (buf))
		fread (buf, 1, skip < BUFSIZ? skip: BUFSIZ, stdin);
	}
    }

    bufp = buf;
    bufcount = fread (bufp, 1, BUFSIZ, stdin);

    while (bufcount > 0) {
	struct disassemble *das;

	if (bufcount < CMD_MAXLEN && ! feof (stdin)) {
	    memcpy (buf, bufp, bufcount);
	    bufp = buf;
	    bufcount += fread (&buf[bufcount], 1, BUFSIZ, stdin);
	    if (ferror (stdin)) {
		perror (infile_n);
		exit (1);
	    }
	}

	das = disassemble (bufp, bufcount);

	printf ("0x%08lx: ", point);
	point += das->size;

	for (i = 0; i < das->size; i++)
	    printf ("%c", isprint (bufp[i])? bufp[i]: '.');
	printf ("%*s", CMD_MAXLEN - i, "");

	printf (" %-23s %s", das->dump, das->mnem);
	if (das->hasarg) {
	    if (das->pcrel)
		printf (" 0x%08lx", point + das->arg);
	    else {
		printf (" %ld", das->arg);
		if (das->arg > 10 || das->arg < -2)
		    printf (" (0x%lx)", das->arg);
	    }
	}
	printf ("\n");

	bufp += das->size;
	bufcount -= das->size;
    }

    return 0;
}
